File: bin/ruby-cli/bin/console-log-remover.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '.common/script_base'

# Script to remove single-line console.log statements from source files
class ConsoleLogRemover < ScriptBase
  def script_emoji; '🧹'; end
  def script_title; 'Console Log Remover'; end
  def script_description; 'Removes single-line console.log statements from source files'; end
  def script_arguments; '[OPTIONS] <extension>'; end

  def add_custom_options(opts)
    opts.on('-d', '--directory DIR', 'Target directory (default: current directory)') do |dir|
      @options[:directory] = dir
    end
    opts.on('-r', '--recursive', 'Search recursively in subdirectories') do
      @options[:recursive] = true
    end
    opts.on('--backup', 'Create backup files before modification') do
      @options[:backup] = true
    end
    opts.on('--dry-run', 'Show what would be removed without making changes') do
      @options[:dry_run] = true
    end
  end

  def validate!
    super

    if args.empty?
      log_error("Missing required argument: file extension")
      log_info("Usage: #{script_name} [OPTIONS] <extension>")
      log_info("Example: #{script_name} js")
      log_info("Example: #{script_name} --directory src --recursive ts")
      exit 1
    end

    @extension = args[0].start_with?('.') ? args[0] : ".#{args[0]}"
    @target_dir = @options[:directory] ? File.expand_path(@options[:directory]) : original_working_dir

    unless File.directory?(@target_dir)
      log_error("Directory not found: #{@target_dir}")
      exit 1
    end
  end

  def run
    log_banner(script_title)
    log_info("Target directory: #{@target_dir}")
    log_info("File extension: #{@extension}")
    log_info("Recursive search: #{@options[:recursive] ? 'Yes' : 'No'}")

    if @options[:backup]
      log_info("Backup files will be created")
    end

    if @options[:dry_run]
      log_warning("DRY RUN MODE - No files will be modified")
    end

    files = find_target_files
    if files.empty?
      log_warning("No files found with extension '#{@extension}'")
      return
    end

    log_info("Found #{files.length} files to process")

    process_files(files)
    show_completion(script_title)
  end

  private

  def find_target_files
    pattern = @options[:recursive] ? "**/*#{@extension}" : "*#{@extension}"

    Dir.chdir(@target_dir) do
      Dir.glob(pattern).select { |f| File.file?(f) }
    end
  end

  def process_files(files)
    stats = { files_processed: 0, logs_removed: 0, files_modified: 0 }

    files.each do |file|
      process_file(file, stats)
    end

    log_section("Summary")
    log_info("Files processed: #{stats[:files_processed]}")
    log_info("Console.log statements removed: #{stats[:logs_removed]}")
    log_info("Files modified: #{stats[:files_modified]}")
  end

  def process_file(file, stats)
    stats[:files_processed] += 1
    file_path = File.join(@target_dir, file)

    begin
      content = File.read(file_path, encoding: 'UTF-8')
      original_lines = content.lines
      modified_lines = []
      logs_removed_this_file = 0

      original_lines.each_with_index do |line, index|
        if console_log_line?(line.strip)
          logs_removed_this_file += 1
          log_debug("Removed console.log from #{file}:#{index + 1}") if debug?
          # Skip this line (don't add to modified_lines)
        else
          modified_lines << line
        end
      end

      if logs_removed_this_file > 0
        stats[:logs_removed] += logs_removed_this_file
        stats[:files_modified] += 1

        log_success("Removed #{logs_removed_this_file} console.log statement(s) from #{file}")

        unless @options[:dry_run]
          # Create backup if requested
          if @options[:backup]
            backup_path = "#{file_path}.backup#{Time.now.strftime('%Y%m%d_%H%M%S')}"
            File.write(backup_path, content)
            log_file_backed_up(backup_path) if verbose?
          end

          # Write modified content
          File.write(file_path, modified_lines.join)
          log_file_updated(file_path) if verbose?
        end
      else
        log_debug("No console.log statements found in #{file}") if debug?
      end

    rescue => e
      log_error("Failed to process #{file}: #{e.message}")
    end
  end

  def console_log_line?(line)
    # Skip empty lines and comments
    return false if line.empty? || line.start_with?('//') || line.start_with?('/*') || line.start_with?('*')

    # Match various console.log patterns
    console_log_patterns = [
      /\A\s*console\.log\s*\(/,                   # console.log(
      /\A\s*console\.debug\s*\(/,                 # console.debug(
      /\A\s*console\.info\s*\(/,                  # console.info(
      /\A\s*console\.warn\s*\(/,                  # console.warn(
      /\A\s*console\.error\s*\(/,                 # console.error(
      /\A\s*console\.log\(`.*`\)\s*\z/,           # console.log(`...`) - template literals
      /\A\s*console\.log\('.*'\)\s*\z/,           # console.log('...') - single quotes
      /\A\s*console\.log\(".*"\)\s*\z/,           # console.log("...") - double quotes
      /\A\s*console\.log\([^)]+\)\s*\z/           # console.log(...) - general pattern
    ]

    # Check if the line contains only console.log statement (possibly with whitespace)
    # and no other code on the same line
    stripped_line = line.strip
    return false unless stripped_line.start_with?('console.')

    # Check if it's a single-line console statement (not multi-line)
    # The line should end with ) or ; after the console call
    return true if console_log_patterns.any? { |pattern| stripped_line.match?(pattern) }

    false
  end

  def show_examples
    puts "Examples:"
    puts "  #{script_name} js                              # Process .js files in current directory"
    puts "  #{script_name} --directory src --recursive ts  # Process .ts files recursively in src/"
    puts "  #{script_name} --backup --dry-run js           # Show what would be removed with backup"
    puts "  #{script_name} --recursive --verbose jsx       # Process .jsx files recursively with verbose output"
  end
end

ConsoleLogRemover.execute if __FILE__ == $0

File: bin/bun-cli/src/scripts/files/console-log-remover.ts
import { Script } from "../../core/decorators/Script";
import { BaseScript } from "../../core/base/Script";
import { logger } from "../../core/utils/logger";
import { readdir, readFile, writeFile, stat } from "node:fs/promises";
import { join, extname } from "node:path";

interface ConsoleLogRemoverOptions {
  directory?: string;
  recursive?: boolean;
  backup?: boolean;
  dryRun?: boolean;
}

/**
 * Removes single-line console.log statements from source files
 */
@Script({
  name: "console-log-remover",
  description: "Removes single-line console.log statements from source files",
  emoji: "🧹",
  arguments: "[OPTIONS] <extension>",
  examples: [
    "console-log-remover js                              # Process .js files in current directory",
    "console-log-remover --directory src --recursive ts  # Process .ts files recursively in src/",
    "console-log-remover --backup --dry-run js           # Show what would be removed with backup",
    "console-log-remover --recursive --verbose jsx       # Process .jsx files recursively with verbose output",
  ],
  options: [
    {
      flag: "-d, --directory <dir>",
      description: "Target directory (default: current directory)",
    },
    {
      flag: "-r, --recursive",
      description: "Search recursively in subdirectories",
    },
    {
      flag: "--backup",
      description: "Create backup files before modification",
    },
    {
      flag: "--dry-run",
      description: "Show what would be removed without making changes",
    },
  ],
})
export class ConsoleLogRemoverScript extends BaseScript {
  private extension!: string;
  private targetDir!: string;
  private options: ConsoleLogRemoverOptions = {};

  async run(args: string[], options: Record<string, any>): Promise<void> {
    logger.banner("Console Log Remover");

    // Parse options
    this.options = {
      directory: options.directory,
      recursive: options.recursive || false,
      backup: options.backup || false,
      dryRun: options.dryRun || false,
    };

    // Validate arguments
    if (args.length === 0) {
      logger.error("Missing required argument: file extension");
      logger.info("Usage: console-log-remover [OPTIONS] <extension>");
      logger.info("Example: console-log-remover js");
      logger.info("Example: console-log-remover --directory src --recursive ts");
      process.exit(1);
    }

    this.extension = args[0].startsWith(".") ? args[0] : `.${args[0]}`;
    this.targetDir = this.options.directory
      ? this.options.directory
      : process.cwd();

    // Verify target directory exists
    try {
      const stats = await stat(this.targetDir);
      if (!stats.isDirectory()) {
        logger.error(`Not a directory: ${this.targetDir}`);
        process.exit(1);
      }
    } catch {
      logger.error(`Directory not found: ${this.targetDir}`);
      process.exit(1);
    }

    logger.info(`Target directory: ${this.targetDir}`);
    logger.info(`File extension: ${this.extension}`);
    logger.info(`Recursive search: ${this.options.recursive ? "Yes" : "No"}`);

    if (this.options.backup) {
      logger.info("Backup files will be created");
    }

    if (this.options.dryRun) {
      logger.warning("DRY RUN MODE - No files will be modified");
    }

    // Find target files
    const files = await this.findTargetFiles(this.targetDir);

    if (files.length === 0) {
      logger.warning(`No files found with extension '${this.extension}'`);
      return;
    }

    logger.info(`Found ${files.length} files to process`);

    // Process files
    await this.processFiles(files);

    logger.completion("Console Log Remover");
  }

  private async findTargetFiles(dir: string): Promise<string[]> {
    const files: string[] = [];

    const entries = await readdir(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = join(dir, entry.name);

      if (entry.isDirectory() && this.options.recursive) {
        const subFiles = await this.findTargetFiles(fullPath);
        files.push(...subFiles);
      } else if (entry.isFile() && extname(entry.name) === this.extension) {
        files.push(fullPath);
      }
    }

    return files;
  }

  private async processFiles(files: string[]): Promise<void> {
    const stats = {
      filesProcessed: 0,
      logsRemoved: 0,
      filesModified: 0,
    };

    for (const file of files) {
      await this.processFile(file, stats);
    }

    logger.section("Summary");
    logger.info(`Files processed: ${stats.filesProcessed}`);
    logger.info(`Console.log statements removed: ${stats.logsRemoved}`);
    logger.info(`Files modified: ${stats.filesModified}`);
  }

  private async processFile(
    filePath: string,
    stats: { filesProcessed: number; logsRemoved: number; filesModified: number }
  ): Promise<void> {
    stats.filesProcessed++;

    try {
      const content = await readFile(filePath, "utf-8");
      const originalLines = content.split("\n");
      const modifiedLines: string[] = [];
      let logsRemovedThisFile = 0;

      for (let i = 0; i < originalLines.length; i++) {
        const line = originalLines[i];
        if (this.isConsoleLogLine(line.trim())) {
          logsRemovedThisFile++;
          logger.debug(
            `Removed console.log from ${filePath}:${i + 1}`,
            this.options.dryRun
          );
          // Skip this line (don't add to modifiedLines)
        } else {
          modifiedLines.push(line);
        }
      }

      if (logsRemovedThisFile > 0) {
        stats.logsRemoved += logsRemovedThisFile;
        stats.filesModified++;

        logger.success(
          `Removed ${logsRemovedThisFile} console.log statement(s) from ${filePath}`
        );

        if (!this.options.dryRun) {
          // Create backup if requested
          if (this.options.backup) {
            const timestamp = new Date()
              .toISOString()
              .replace(/[-:]/g, "")
              .replace(/\..+/, "");
            const backupPath = `${filePath}.backup${timestamp}`;
            await writeFile(backupPath, content);
            logger.debug(`Backed up to: ${backupPath}`, true);
          }

          // Write modified content
          await writeFile(filePath, modifiedLines.join("\n"));
          logger.debug(`Updated: ${filePath}`, true);
        }
      } else {
        logger.debug(`No console.log statements found in ${filePath}`, false);
      }
    } catch (error) {
      logger.error(
        `Failed to process ${filePath}: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }

  private isConsoleLogLine(line: string): boolean {
    // Skip empty lines and comments
    if (
      !line ||
      line.startsWith("//") ||
      line.startsWith("/*") ||
      line.startsWith("*")
    ) {
      return false;
    }

    // Must start with console.
    if (!line.startsWith("console.")) {
      return false;
    }

    // Match various console.log patterns
    const consoleLogPatterns = [
      /^\s*console\.log\s*\(/,              // console.log(
      /^\s*console\.debug\s*\(/,            // console.debug(
      /^\s*console\.info\s*\(/,             // console.info(
      /^\s*console\.warn\s*\(/,             // console.warn(
      /^\s*console\.error\s*\(/,            // console.error(
      /^\s*console\.log\(`.*`\)\s*$/,       // console.log(`...`) - template literals
      /^\s*console\.log\('.*'\)\s*$/,       // console.log('...') - single quotes
      /^\s*console\.log\(".*"\)\s*$/,       // console.log("...") - double quotes
      /^\s*console\.log\([^)]+\)\s*$/,      // console.log(...) - general pattern
    ];

    // Check if the line matches any console.log pattern
    return consoleLogPatterns.some((pattern) => pattern.test(line));
  }
}
